core/services/gas_updater.go

```diff
@@ -57,91 +70,292 @@ type gasUpdater struct {
-	if !gu.store.Config.GasUpdaterEnabled() {
-		return
+	if gu.State() != utils.StartStopOnce_Started {
+		panic("OnNewLongestChain called on unstarted gas updater")
+	}
+	gu.mb.Deliver(head)
+}
+
+func (gu *gasUpdater) Start() error {
+	if !gu.OkayToStart() {
+		return errors.New("GasUpdater has already been started")
+	}
+	gu.logger.Debugw("GasUpdater: starting")
+	if gu.rollingBlockHistorySize > int64(gu.store.Config.EthFinalityDepth()) {
+		gu.logger.Warnf("GasUpdater: GAS_UPDATER_BLOCK_HISTORY_SIZE=%v is greater than ETH_FINALITY_DEPTH=%v, blocks deeper than finality depth will be refetched on every gas updater cycle, causing unnecessary load on the eth node. Consider decreasing GAS_UPDATER_BLOCK_HISTORY_SIZE or increasing ETH_FINALITY_DEPTH", gu.rollingBlockHistorySize, gu.store.Config.EthFinalityDepth())
+	}
+	ctx, cancel := context.WithTimeout(gu.ctx, maxStartTime)
+	defer cancel()
+	latestHead, err := gu.store.EthClient.HeaderByNumber(ctx, nil)
+	if err != nil {
+		logger.Warnw("GasUpdater: initial check for latest head failed", "err", err)
+	} else {
+		gu.logger.Debugw("GasUpdater: got latest head", "number", latestHead.Number, "blockHash", latestHead.Hash.Hex())
+		gu.FetchBlocksAndRecalculate(ctx, *latestHead)
+	}
+	gu.wg.Add(1)
+	go gu.runLoop()
+	gu.logger.Debugw("GasUpdater: started")
+	return nil
+}
+
+func (gu *gasUpdater) Close() error {
+	if !gu.OkayToStop() {
+		return errors.New("GasUpdater has already been stopped")
 	}
+	gu.ctxCancel()
+	gu.wg.Wait()
+	return nil
+}
+
+func (gu *gasUpdater) runLoop() {
+	defer gu.wg.Done()
+	for {
+		select {
+		case <-gu.ctx.Done():
+			return
+		case <-gu.mb.Notify():
+			head := gu.mb.Retrieve()
+			h, is := head.(models.Head)
+			if !is {
+				panic(fmt.Sprintf("invariant violation, expected %T but got %T", models.Head{}, head))
+			}
+			gu.FetchBlocksAndRecalculate(gu.ctx, h)
+		}
+	}
+}
+
+func (gu *gasUpdater) FetchBlocksAndRecalculate(ctx context.Context, head models.Head) {
 	ctx, cancel := context.WithTimeout(ctx, maxEthNodeRequestTime)
 	defer cancel()
-	blockToFetch := head.Number - gu.blockDelay
-	if blockToFetch < 0 {
-		logger.Warnf("GasUpdater: skipping gas calculation, current block height %v is lower than GAS_UPDATER_BLOCK_DELAY of %v", head.Number, gu.blockDelay)
+
+	if err := gu.FetchBlocks(ctx, head); err != nil {
+		gu.logger.Warnw("GasUpdater: error fetching blocks", "head", head, "err", err)
+		return
+	}
+
+	gu.Recalculate(head)
+}
+
+// FetchHeadsAndRecalculate adds the given heads to the history and recalculates gas price
+func (gu *gasUpdater) Recalculate(head models.Head) {
+	if len(gu.rollingBlockHistory) == 0 {
 		return
 	}
-	block, err := gu.store.EthClient.BlockByNumber(ctx, big.NewInt(blockToFetch))
+
+	percentileGasPrice, err := gu.percentileGasPrice()
 	if err != nil {
-		logger.Errorf("GasUpdater: error retrieving block %v: %s", blockToFetch, err)
+		logger.Errorw("GasUpdater: error getting percentile gas price", "err", err)
```

Comment:
I don't think we need to wake someone up if there was a string empty blocks, it'd still use the last updated gasPrice which would be accurate. Can make this a warn